--- libtorrent-0.15.5/src/torrent/peer/client_list.cc~	2025-07-07 14:23:46.000000000 +0200
+++ libtorrent-0.15.5/src/torrent/peer/client_list.cc	2025-07-07 14:28:54.586475195 +0200
@@ -60,7 +60,7 @@ ClientList::ClientList() {
   insert_helper(ClientInfo::TYPE_AZUREUS, "qB", NULL, NULL, "qBittorrent");
   insert_helper(ClientInfo::TYPE_AZUREUS, "UT", NULL, NULL, "uTorrent");
   insert_helper(ClientInfo::TYPE_AZUREUS, "TR", NULL, NULL, "Transmission");
-  insert_helper(ClientInfo::TYPE_AZUREUS, "DE", NULL, NULL, "DelugeTorrent");
+  insert_helper(ClientInfo::TYPE_AZUREUS, "DE", NULL, NULL, "Deluge");
   insert_helper(ClientInfo::TYPE_AZUREUS, "AZ", NULL, NULL, "Vuze");
   insert_helper(ClientInfo::TYPE_AZUREUS, "UM", NULL, NULL, "uTorrent Mac");
   insert_helper(ClientInfo::TYPE_AZUREUS, "LT", NULL, NULL, "libtorrent");
@@ -70,6 +70,7 @@ ClientList::ClientList() {
   insert_helper(ClientInfo::TYPE_AZUREUS, "BC", NULL, NULL, "BitComet");
   insert_helper(ClientInfo::TYPE_AZUREUS, "XL", NULL, NULL, "Xunlei");
   insert_helper(ClientInfo::TYPE_AZUREUS, "SD", NULL, NULL, "Xunlei");
+  insert_helper(ClientInfo::TYPE_AZUREUS, "XX", NULL, NULL, "Xtorrent");
 
   // Other clients.
   insert_helper(ClientInfo::TYPE_AZUREUS, "7T", NULL, NULL, "aTorrent");
@@ -146,7 +147,7 @@ ClientList::ClientList() {
   insert_helper(ClientInfo::TYPE_AZUREUS, "RS", NULL, NULL, "Rufus");
   insert_helper(ClientInfo::TYPE_AZUREUS, "RT", NULL, NULL, "Retriever");
   insert_helper(ClientInfo::TYPE_AZUREUS, "RZ", NULL, NULL, "RezTorrent");
-  insert_helper(ClientInfo::TYPE_COMPACT, "S", NULL, NULL, "Shadow's client");
+  insert_helper(ClientInfo::TYPE_COMPACT, "S", NULL, NULL, "Shadow's");
   insert_helper(ClientInfo::TYPE_AZUREUS, "S~", NULL, NULL, "Shareaza alpha/beta");
   insert_helper(ClientInfo::TYPE_AZUREUS, "SB", NULL, NULL, "SwiftBit");
   insert_helper(ClientInfo::TYPE_AZUREUS, "SG", NULL, NULL, "GS Torrent");
@@ -171,6 +177,13 @@ ClientList::retrieve_id(ClientInfo* dest
       std::isxdigit(id[3]) && std::isxdigit(id[4]) && std::isxdigit(id[5]) && std::isxdigit(id[6])) {
     dest->set_type(ClientInfo::TYPE_AZUREUS);
 
+    // incoming connection with spoofed id
+    if (id[1] == 'l' && id[2] == 't' && !std::strncmp(id.c_str()+16, "UDP0", 4)) {
+      *dest = *begin();
+      std::memset(dest->mutable_upper_version(), 0, ClientInfo::max_version_size);
+      return false;
+    }
+
     dest->mutable_key()[0] = id[1];
     dest->mutable_key()[1] = id[2];
     
@@ -178,15 +191,22 @@ ClientList::retrieve_id(ClientInfo* dest
       dest->mutable_version()[i] = dest->mutable_upper_version()[i] = rak::hexchar_to_value(id[3 + i]);
 
   } else if (std::isalpha(id[0]) && id[4] == '-' &&
-             std::isxdigit(id[1]) && std::isxdigit(id[2]) && std::isxdigit(id[3])) {
+             std::isxdigit(id[1]) && std::isxdigit(id[2]) && std::isalnum(id[3])) {
     dest->set_type(ClientInfo::TYPE_COMPACT);
 
     dest->mutable_key()[0] = id[0];
     dest->mutable_key()[1] = '\0';
     
+    // T03G = BitTornado 0.3.16
+    char id3;
+    if (id[3] > 'F' && id[3] < 'a')
+      id3 = id[3] - 'F' + 0xF;
+    else
+      id3 = rak::hexchar_to_value(id[3]);
+
     dest->mutable_version()[0] = dest->mutable_upper_version()[0] = rak::hexchar_to_value(id[1]);
     dest->mutable_version()[1] = dest->mutable_upper_version()[1] = rak::hexchar_to_value(id[2]);
-    dest->mutable_version()[2] = dest->mutable_upper_version()[2] = rak::hexchar_to_value(id[3]);
+    dest->mutable_version()[2] = dest->mutable_upper_version()[2] = id3;
     dest->mutable_version()[3] = dest->mutable_upper_version()[3] = '\0';
 
   } else if (std::isalpha(id[0]) && std::isdigit(id[1]) && id[2] == '-' &&
