diff --git a/src/tracker/tracker_http.cc b/src/tracker/tracker_http.cc
index 6dbd0f5..48ab5a5 100644
--- a/src/tracker/tracker_http.cc
+++ b/src/tracker/tracker_http.cc
@@ -276,8 +276,11 @@ TrackerHttp::receive_done() {
   Object b;
   *m_data >> b;
 
-  if (m_data->fail())
-    return receive_failed("Could not parse bencoded data");
+  if (m_data->fail()) {
+    std::stringstream dump;
+    dump << "Could not parse bencoded data: " << m_data->str();
+    return receive_failed(dump.str().c_str());
+  }
 
   if (!b.is_map())
     return receive_failed("Root not a bencoded map");
